File: src/building/BuildingSystem.js
/**
 * Building System
 * Manages building placement and construction
 */
class BuildingSystem {
    constructor(world, audioManager = null, assetLoader = null) {
        this.world = world;
        this.audioManager = audioManager;
        this.assetLoader = assetLoader;
        this.buildMode = false;
        this.selectedBuildingType = Building.TYPES.HOUSE;
        this.previewTile = null;
    }
    
    /**
     * Toggle build mode
     */
    toggleBuildMode() {
        this.buildMode = !this.buildMode;
        return this.buildMode;
    }
    
    /**
     * Set building type to place
     */
    setBuildingType(type) {
        this.selectedBuildingType = type;
    }
    
    /**
     * Update preview based on mouse position
     */
    updatePreview(mouseX, mouseY, camera) {
        if (!this.buildMode) {
            this.previewTile = null;
            return;
        }
        
        const tile = this.world.getTileAtScreen(mouseX, mouseY, camera);
        this.previewTile = tile;
    }
    
    /**
     * Check if building can be placed at tile
     */
    canPlaceBuilding(tile) {
        if (!tile) return false;
        if (!tile.isWalkable()) return false;
        if (tile.building) return false;
        
        // Check if it's a valid building location
        const buildingWidth = this.selectedBuildingType.width;
        const buildingHeight = this.selectedBuildingType.height;
        
        // Check all tiles that building will occupy
        for (let dy = 0; dy < buildingHeight; dy++) {
            for (let dx = 0; dx < buildingWidth; dx++) {
                const checkTile = this.world.getTile(tile.x + dx, tile.y + dy);
                if (!checkTile || !checkTile.isWalkable() || checkTile.building) {
                    return false;
                }
            }
        }
        
        return true;
    }
    
    /**
     * Place building at tile
     */
    placeBuilding(tile) {
        if (!this.canPlaceBuilding(tile)) {
            // Play error sound if available
            if (this.audioManager) {
                this.audioManager.playSfx('sfx_womp', 0.5);
            }
            return false;
        }
        
        const building = new Building(tile.x, tile.y, this.selectedBuildingType, this.assetLoader);
        
        // Occupy all tiles
        const buildingWidth = this.selectedBuildingType.width;
        const buildingHeight = this.selectedBuildingType.height;
        
        for (let dy = 0; dy < buildingHeight; dy++) {
            for (let dx = 0; dx < buildingWidth; dx++) {
                const occupyTile = this.world.getTile(tile.x + dx, tile.y + dy);
                if (occupyTile) {
                    occupyTile.setBuilding(building);
                }
            }
        }
        
        // Play placement sound if available
        if (this.audioManager) {
            this.audioManager.playSfx('sfx_badadadink', 0.7);
        }
        
        return true;
    }
    
    /**
     * Attempt to place building at mouse position
     */
    tryPlaceBuilding(mouseX, mouseY, camera) {
        if (!this.buildMode) return false;
        
        const tile = this.world.getTileAtScreen(mouseX, mouseY, camera);
        return this.placeBuilding(tile);
    }
    
    /**
     * Render building preview
     */
    renderPreview(renderer, camera, isometricRenderer) {
        if (!this.buildMode || !this.previewTile) return;
        
        const canPlace = this.canPlaceBuilding(this.previewTile);
        const color = canPlace ? 'rgba(0, 255, 0, 0.3)' : 'rgba(255, 0, 0, 0.3)';
        
        const screenPos = IsometricUtils.tileToScreen(
            this.previewTile.x,
            this.previewTile.y,
            64,
            32
        );
        
        // Draw preview cube
        renderer.save();
        renderer.setAlpha(0.5);
        
        isometricRenderer.drawIsometricCube(
            screenPos.x,
            screenPos.y,
            64,
            32,
            this.selectedBuildingType.buildHeight,
            color,
            color,
            color,
            camera
        );
        
        renderer.restore();
        
        // Draw outline
        isometricRenderer.drawIsometricTileOutline(
            screenPos.x,
            screenPos.y,
            64,
            32,
            canPlace ? '#00ff00' : '#ff0000',
            camera,
            2
        );
    }
}

File: cpp/src/building/BuildingSystem.cpp
#include "building/BuildingSystem.h"
#include "world/World.h"
#include "rendering/Renderer.h"
#include "rendering/IsometricRenderer.h"
#include <iostream>

BuildingSystem::BuildingSystem(World* world)
    : world(world)
{
}

void BuildingSystem::update(float deltaTime) {
    // Building system update logic (future: construction progress, etc.)
    (void)deltaTime; // Unused - reserved for future construction progress
}

void BuildingSystem::render(Renderer* renderer, IsometricRenderer* isoRenderer, Camera* camera) {
    (void)renderer; // Unused - using isoRenderer for rendering
    (void)camera; // Unused - camera handled by renderer
    // Render all buildings
    for (const auto& building : buildings) {
        isoRenderer->drawIsometricCube(
            building->getX(),
            building->getY(),
            building->getBuildHeight(),
            building->getTopColor(),
            building->getLeftColor(),
            building->getRightColor()
        );
    }
}

bool BuildingSystem::placeBuilding(int x, int y, BuildingType type) {
    // Check if building can be placed
    if (!canPlaceBuilding(x, y, type)) {
        return false;
    }
    
    // Create building
    auto building = std::make_unique<Building>(x, y, type);
    
    // Mark tiles as occupied
    markTilesOccupied(x, y, building->getWidth(), building->getHeight(), true);
    
    buildings.push_back(std::move(building));
    
    std::cout << "Placed building at (" << x << ", " << y << ")" << std::endl;
    return true;
}

bool BuildingSystem::canPlaceBuilding(int x, int y, BuildingType type) const {
    // Get building dimensions based on type
    int width, height;
    
    switch (type) {
        case BuildingType::HOUSE:
            width = 2; height = 2;
            break;
        case BuildingType::TOWER:
            width = 1; height = 1;
            break;
        case BuildingType::WAREHOUSE:
            width = 3; height = 3;
            break;
        default:
            return false;
    }
    
    return areTilesAvailable(x, y, width, height);
}

Building* BuildingSystem::getBuildingAt(int x, int y) {
    for (auto& building : buildings) {
        int bx = building->getX();
        int by = building->getY();
        int bw = building->getWidth();
        int bh = building->getHeight();
        
        if (x >= bx && x < bx + bw && y >= by && y < by + bh) {
            return building.get();
        }
    }
    return nullptr;
}

bool BuildingSystem::removeBuilding(int x, int y) {
    for (auto it = buildings.begin(); it != buildings.end(); ++it) {
        if ((*it)->getX() == x && (*it)->getY() == y) {
            // Mark tiles as unoccupied
            markTilesOccupied(x, y, (*it)->getWidth(), (*it)->getHeight(), false);
            
            buildings.erase(it);
            std::cout << "Removed building at (" << x << ", " << y << ")" << std::endl;
            return true;
        }
    }
    return false;
}

bool BuildingSystem::areTilesAvailable(int x, int y, int width, int height) const {
    for (int dy = 0; dy < height; ++dy) {
        for (int dx = 0; dx < width; ++dx) {
            const Tile* tile = world->getTile(x + dx, y + dy);
            
            if (!tile) {
                return false; // Out of bounds
            }
            
            if (!tile->isWalkable() || tile->isOccupied()) {
                return false; // Tile not available
            }
        }
    }
    return true;
}

void BuildingSystem::markTilesOccupied(int x, int y, int width, int height, bool occupied) {
    for (int dy = 0; dy < height; ++dy) {
        for (int dx = 0; dx < width; ++dx) {
            Tile* tile = world->getTile(x + dx, y + dy);
            if (tile) {
                tile->setOccupied(occupied);
            }
        }
    }
}
